mne_bids/utils.py

```diff
@@ -621,3 +622,75 @@ def _find_matching_sidecar(bids_fname, bids_root, suffix, allow_fail=False):
         return None
     else:
         raise RuntimeError(msg)
+
+
+def _update_sidecar(sidecar_fname, key, val):
+    """Update a sidecar JSON file with a given key/value pair.
+
+    Parameters
+    ----------
+    sidecar_fname : str
+        Full name of the data file
+    key : str
+        The key in the sidecar JSON file. E.g. "PowerLineFrequency"
+    val : str
+        The corresponding value to change to in the sidecar JSON file.
+    """
+    with open(sidecar_fname, "r") as fin:
+        sidecar_json = json.load(fin)
+    sidecar_json[key] = val
+    with open(sidecar_fname, "w") as fout:
+        json.dump(sidecar_json, fout)
+
+
+def _estimate_line_freq(raw, verbose=False):
+    """Estimate power line noise from a given BaseRaw.
+
+    Parameters
+    ----------
+    raw : mne.io.BaseRaw
+
+    Returns
+    -------
+    line_freq : int
+        Either 50, or 60 Hz depending if European,
+        or USA data recording.
+    """
+    sfreq = raw.info['sfreq']
+
+    # setup picks of the data
+    picks = pick_types(raw.info, meg=True, eeg=True,
+                       stim=True, ecog=True, seeg=True,
+                       exclude='bads')
+
+    # only sample first 10 seconds, or whole time series
+    tmin = 0
+    tmax = int(min(len(raw.times), 10 * sfreq))
+
+    # get just five channels of data to estimate on
+    data = raw.get_data(start=tmin, stop=tmax,
+                        picks=picks, return_times=False)[0:5, :]
+
+    # if sampling is not high enough, line_freq does not matter
+    if sfreq < 100:
+        return None
+
+    # run a multi-taper FFT between Power Line Frequencies of interest
+    psds, freqs = psd_array_welch(data, fmin=49, fmax=61,
+                                  sfreq=sfreq, average="mean")
+    usa_ind = np.where(freqs == min(freqs, key=lambda x: abs(x - 60)))[0]
+    eu_ind = np.where(freqs == min(freqs, key=lambda x: abs(x - 50)))[0]
+
+    # get the average power within those frequency bands
+    usa_psd = np.mean((psds[..., usa_ind]))
+    eu_psd = np.mean((psds[..., eu_ind]))
```

Comment:
as you have data from different channels types you may end up averaging V with T or T/m. Not sure I have a better idea but can it be a problem in this context?

Reply:
I set it up now to only estimate for meg, eeg, ecog, or seeg for sampling above 100 Hz. It uses an estimate that combines data from at at least 5 channels of one acquisition modality.